modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -215,15 +215,30 @@ private void analyzeCompound(final CompilerSettings settings, final Definition d
         there = AnalyzerCaster.getLegalCast(definition, location, last.after, promote, false);
         back = AnalyzerCaster.getLegalCast(definition, location, promote, last.after, true);
 
+        if (last instanceof ADefLink) {
+            final ADefLink lastDef = (ADefLink) last;
+            // Unfortunately, we don't know the real type because we load from DEF and store to DEF!
+            lastDef.storeValueType = last.after;
+        }
+
         statement = true;
         actual = read ? last.after : definition.voidType;
     }
 
     private void analyzeWrite(final CompilerSettings settings, final Definition definition, final Variables variables) {
         final ALink last = links.get(links.size() - 1);
 
-        expression.expected = last.after;
-        expression.analyze(settings, definition, variables);
+        // If the store node is a DEF node, we remove the cast to DEF from the expression
+        // and promote the real type to it:
+        if (last instanceof ADefLink) {
+            final ADefLink lastDef = (ADefLink) last;
+            expression.analyze(settings, definition, variables);
+            lastDef.storeValueType = expression.expected = expression.actual;
```

Comment:
I guess I'm a bit confused on how this is working because to me it looks like when this instruction is executed ```INVOKESTATIC org/elasticsearch/painless/Def.DefToint (Ljava/lang/Object;)I``` there is an int value on the stack, not an object?